agent/lib/kontena/websocket_client.rb

```diff
@@ -62,142 +55,235 @@ def connecting?
+      defer {
+        ws.on_message do |message|
+          actor.on_message(message)
+        end
+        ws.on_pong do |delay|
+          actor.on_pong(delay)
+        end
+
+        # raises on errors
+        ws.run do
+          # These are called from the read_ws -> defer thread, proxy back to actor
+          actor.on_open
         end
       }
+
+    rescue Kontena::Websocket::Error => exc
+      # handle known errors, will reconnect
+      on_error exc
+
     rescue => exc
-      error "failed to send message: #{exc.message}"
+      # XXX: crash instead of reconnecting on unknown errors?
+      error exc
+
+    else
+      on_close(ws.close_code, ws.close_reason)
+
+    ensure
+      @connected = false
+      @connecting = false
+      @ws = nil
     end
 
+    def ws
+      fail "not connected" unless @ws
+
+      @ws
+    end
+
+    # Called from RpcServer, does not crash the Actor on errors.
+    #
+    # @param [String, Array] msg
+    # @raise [RuntimeError] not connected
+    def send_message(msg)
+      ws.send(msg)
+    rescue => exc
+      warn exc
+      abort exc
+    end
+
+    # Called from RpcClient, does not crash the Actor on errors.
+    #
     # @param [String] method
     # @param [Array] params
+    # @raise [RuntimeError] not connected
     def send_notification(method, params)
       data = MessagePack.dump([2, method, params]).bytes
-      send_message(data)
+      ws.send(data)
     rescue => exc
-      error "failed to send notification: #{exc.message}"
+      warn exc
+      abort exc
     end
 
+    # Called from RpcClient, does not crash the Actor on errors.
+    #
     # @param [Integer] id
     # @param [String] method
     # @param [Array] params
+    # @raise [RuntimeError] not connected
     def send_request(id, method, params)
       data = MessagePack.dump([0, id, method, params]).bytes
-      send_message(data)
+      ws.send(data)
     rescue => exc
-      error "failed to send request: #{exc.message}"
+      warn exc
+      abort exc
     end
 
-    # @param [Faye::WebSocket::API::Event] event
-    def on_open(event)
-      ping_timer.cancel if ping_timer
-      info 'connection established'
+    def on_open
       @connected = true
       @connecting = false
-      notify_actors('websocket:open', event)
+
+      ssl_verify = ws.ssl_verify?
+
+      begin
+        ssl_cert = @ws.ssl_cert!
+        ssl_error = nil
+      rescue Kontena::Websocket::SSLVerifyError => exc
+        ssl_cert = exc.cert
+        ssl_error = exc
+      end
+
+      if ssl_error
+        if ssl_cert
+          warn "insecure connection established with SSL errors: #{ssl_error}: #{ssl_cert.subject} (issuer #{ssl_cert.issuer})"
+        else
+          warn "insecure connection established with SSL errors: #{ssl_error}"
+        end
+      elsif ssl_cert
+        if !ssl_verify
+          warn "secure connection established without KONTENA_SSL_VERIFY=true: #{ssl_cert.subject} (issuer #{ssl_cert.issuer})"
```

Comment:
Warning because while this specific connection is secure, the configuration is vulnerable to MITM errors, because the agent will still accept an invalid cert unless you update the configuration to use `KONTENA_SSL_VERIFY=true`.